File: dsl-api/src/main/java/cbs/dsl/builder/WorkflowBuilder.java
package cbs.dsl.builder;

import cbs.dsl.api.Action;
import cbs.dsl.api.DslObject;
import cbs.dsl.api.EventDefinition;
import cbs.dsl.api.ParameterDefinition;
import cbs.dsl.api.TransitionRuleDefinition;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

/** Builder for creating workflow objects from DSL files. */
public class WorkflowBuilder {

  private final String code;
  private final List<String> states = new ArrayList<>();
  private String initialState;
  private final List<String> terminalStates = new ArrayList<>();
  private final List<TransitionRuleDefinition> transitions = new ArrayList<>();

  WorkflowBuilder(String code) {
    this.code = code;
  }

  public WorkflowBuilder states(String... states) {
    Collections.addAll(this.states, states);
    return this;
  }

  public WorkflowBuilder initial(String initial) {
    this.initialState = initial;
    return this;
  }

  public WorkflowBuilder terminal(String... states) {
    Collections.addAll(this.terminalStates, states);
    return this;
  }

  public WorkflowBuilder transition(String from, String to, Action action, EventDefinition event) {
    this.transitions.add(new TransitionRuleDefinition(from, to, action, event));
    return this;
  }

  @Deprecated(forRemoval = true)
  public WorkflowBuilder transition(String from, String to, Action action, String eventCode) {
    EventDefinition proxy = new EventDefinition() {
      @Override
      public String getCode() {
        return eventCode;
      }

      @Override
      public List<ParameterDefinition> getParameters() {
        return Collections.emptyList();
      }
    };
    this.transitions.add(new TransitionRuleDefinition(from, to, action, proxy));
    return this;
  }

  public String getCode() {
    return code;
  }

  public List<String> states() {
    return Collections.unmodifiableList(new ArrayList<>(states));
  }

  public String initial() {
    return initialState;
  }

  public List<String> terminalStates() {
    return Collections.unmodifiableList(new ArrayList<>(terminalStates));
  }

  public List<TransitionRuleDefinition> transitions() {
    return Collections.unmodifiableList(new ArrayList<>(transitions));
  }

  public DslObject build() {
    return WorkflowDslObject.builder()
        .code(code)
        .states(Collections.unmodifiableList(new ArrayList<>(states)))
        .initial(initialState)
        .terminalStates(Collections.unmodifiableList(new ArrayList<>(terminalStates)))
        .transitions(Collections.unmodifiableList(new ArrayList<>(transitions)))
        .build();
  }
}

File: dsl/src/main/kotlin/cbs/dsl/runtime/WorkflowBuilder.kt
package cbs.dsl.runtime

import cbs.dsl.api.Action
import cbs.dsl.api.EventDefinition
import cbs.dsl.api.TransitionRule
import cbs.dsl.api.WorkflowDefinition

class WorkflowBuilder(override val code: String) : WorkflowDefinition {
  private val _states = mutableListOf<String>()
  private var _initial: String = ""
  private val _terminalStates = mutableListOf<String>()
  private val _transitions = mutableListOf<TransitionRule>()

  override val transitions: List<TransitionRule>
    get() = _transitions.toList()

  fun states(vararg s: String) {
    _states.addAll(s)
  }

  fun initial(s: String) {
    _initial = s
  }

  fun terminalStates(vararg s: String) {
    _terminalStates.addAll(s)
  }

  fun transitions(block: TransitionScope.() -> Unit) {
    val scope = TransitionScope()
    scope.block()
    _transitions += scope.rules
  }

  @Deprecated(
      message = "Use transitions { } block with closure syntax instead",
      level = DeprecationLevel.WARNING,
  )
  fun transition(
      from: String,
      to: String,
      on: Action,
      event: EventDefinition,
      onFault: String = "FAULTED",
  ) {
    _transitions += TransitionRule(from, to, on, event, onFault)
  }

  // Auto-inference methods
  private fun inferredStates(): List<String> {
    if (_states.isNotEmpty()) return _states
    return (_transitions.map { it.from } + _transitions.map { it.to }).distinct()
  }

  private fun inferredInitial(): String {
    if (_initial.isNotEmpty()) return _initial
    val targets = _transitions.map { it.to }.toSet()
    return inferredStates().firstOrNull { it !in targets } ?: ""
  }

  private fun inferredTerminal(): List<String> {
    if (_terminalStates.isNotEmpty()) return _terminalStates
    val sources = _transitions.map { it.from }.toSet()
    return inferredStates().filter { it !in sources }
  }

  override val states: List<String>
    get() = inferredStates()

  override val initial: String
    get() = inferredInitial()

  override val terminalStates: List<String>
    get() = inferredTerminal()
}

fun workflow(code: String, block: WorkflowBuilder.() -> Unit): WorkflowDefinition =
    WorkflowBuilder(code).apply(block)
